### ISSUE TITLE
Summary?

### ISSUE THREAD

**Issue by timholy** [eschnett/FixedPoint.jl#1]
Greetings @eschnett. I stumbled across this because of a GitHub search triggered by https://github.com/JuliaMath/FixedPointNumbers.jl/issues/128. This seems potentially interesting. Can you summarize the main goals? Performance, I assume?

**Comment by eschnett** [comment]
Yes, the main goal is to provide high-precision "floating point" (or maybe only fixed-point) numbers. Via macros or generated functions I want to avoid memory allocation overhead that would be present for `BigInt` or `BigFloat`. This is probably only suitable for less than about 1000 bits of precision.

The direct competitor for this package would be `DoubleFloats.jl` and friends, where it seems non-trivial to achieve higher precision by combining more than 2 `Float64` values. In principle there should be a straightforward, composable way to create e.g. `QuadFloat` from `DoubleFloat`, but if I look at the implementation, it's quite messy with lots of awkward helper functions. Hence my experiment based on rather clean integer arithmetic.

Unfortunately I had to discover that modern CPUs have an incredible amount of floating-point performance (in particular for multiplications) and only a rather small integer multiply performance. For example, Intel's SIMD integer multiplications are limited to 32 bits, and building a 64-bit multiplication out of these is quite slow. It's faster to use a scalar 64-bit integer multiplication, even with AVX512. Given this, I don't have much hope that integer-based fixed-point arithmetic can be competitive on current hardware.

I thought I had a comment in this direction in the README, but apparently I don't...

**Comment by timholy** [comment]
If you have/find performance improvements in stuff related to FixedPointNumbers, we're certainly all ears. It hasn't gotten any benchmarking/optimization in a while. But agreed it's a hard problem. Thanks for the summary!
